### ISSUE TITLE
Feature Request: Firefox Port

### ISSUE THREAD

**Issue by eberhardweber** [salcido/discogs-enhancer#105]
Firefox uses the same WebExtension standard as Google Chrome with minor differences, so porting and publishing - while a considerable task - wouldn't mean rewriting the extension completely in order to support a new platform. 

Adding Firefox support would also help with using Discogs on Android phones, as Firefox supports add-ons on the mobile version. 

I currently have to use an installation of Vivaldi just for this add-on so Firefox support would be fantastic! :)

**Comment by salcido** [comment]
@eberhardweber 
Yes, Firefox uses the same standard as Chrome but there are enough differences to make the task of creating a separate version laborious with little to no ROI for myself. 

I tried for several months back in 2016 to get Discogs Enhancer into the Firefox Add-on store. The code reviewers were largely unhelpful, refused to answer my questions, and repeatedly sent me deprecated documentation, while simultaneously requesting I update their outdated documents myself. The process was quite frustrating and left me with little interest in developing extensions on the Firefox platform.

Even when ignoring the reasons above, there are other issues preventing Discogs Enhancer from being accepted into the Add-On store. Mozilla will not allow extensions to use the `innerHTML` method. That means large portions of the codebase [(this for example)](https://github.com/salcido/discogs-enhancer/blob/master/js/extension/features/currency-converter.js#L26-L87) would need to be rewritten using `document.createElement` to generate the necessary markup. This approach makes maintaining, augmenting, and debugging the extension significantly more tedious. 

Another big factor is time. I've been working on this extension for close to 4 years now and it's grown quite a bit. It takes a large amount of time and energy to put out new versions. Anytime there is a new feature I need to:
- write the logic
- test it
- write unit and functional tests for it
- update the readme
- update the change log
- create the images and documentation for the learn page
- update the toggle and submenu logic in the popup
- update the website
- update the list of features on the Chrome store page
- make a new release here on Github
- make an announcement post in the group on Discogs.com
 
Adding Firefox into the mix means even more time doing the aforementioned list of tasks with each new release.

Money also plays a role. Discogs Enhancer uses its own currency conversion API that costs me $72/year to host. The extension is free but I ask for donations. Occasionally generous people will donate a few dollars which helps to offset the cost of hosting. Unfortunately, I only receive around $35 dollars a year in donations. Essentially about enough to cover half of the cost of hosting the API. 

I love Discogs and I enjoy working on this extension but the amount of time it takes to maintain it is substantial and I don't receive enough in donations to make the additional time and energy required to support a Firefox version worth it.

**Comment by eberhardweber** [comment]
Appreciate the definitive answer. Good luck with maintaining the add-on! Disheartened to hear you received poor support. Unfortunately, the Firefox add-ons WebExtension conversion seems to have been coupled with a fair bit of unfortunate streamlining. It often feels to me like add-on developers are seen by the corporate as a nuisance, or a pest, instead of a positive, transformative force.

**Comment by salcido** [comment]
Hey thanks! I haven't looked into this in a number of years but I'll take another look and see how feasible it is.

**Comment by nd-** [comment]
Oh, that would be so cool :)

**Comment by Madslaw** [comment]
As mentioned by slatkin, I don't think it's as tedious as you might think. I currently have a working version of the extension on my Firefox browser. I was able to do it by adapting the format of manifest.json so that it's compatible with the one use by Firefox.

I did it in early June for version 3.9.3, and now with all the changes made to the UI of Discogs, I was able to make it work for version 3.9.4 by transferring my version of the manifest.json to your updated source code.

After that, I added it to Firefox by doing the following steps (which I don't think are very different for Chrome): 
- Compiled the source code with npm run build
- Saved it as a .zip file
- Changed the extension to .xpi
- Submitted it via [https://addons.mozilla.org/en-US/developers/](url) (I did not make my version publicly available since I made it for my own personal use)

Performance wise, everything seems to be working just fine. So far, I have not noticed any bugs. 

I've attatched bellow my version of the manifest.json if you want to reuse it for a possible official Firefox release. Just note that the id in my file "discogs-enhancer@example.com" was just to make it work, and you might want to change it to something else.

[manifest.json](https://github.com/user-attachments/files/20961990/manifest.json)

```
{
  "manifest_version": 3,
  "name": "Discogs Enhancer",
  "short_name": "Discogs Enhancer",
  "description": "Enhance your Discogs experience with dark themes, seller tools, price comparisons, currency conversion, powerful filters, and more!",
  "version": "3.9.4",
  "author": "Matthew Salcido",
  "homepage_url": "https://www.discogs-enhancer.com",
  "browser_specific_settings": {
  "gecko": {
    "id": "discogs-enhancer@example.com"
    }
  },

  "action": {
    "default_icon": {
      "16": "img/icon_16.png",
      "32": "img/icon_32.png",
      "48": "img/icon_48.png",
      "128": "img/icon_128.png"
    },
    "default_title": "Discogs Enhancer",
    "default_popup": "html/popup.html"
  },

  "background": {
    "scripts": ["js/extension/background.js"]
  },

  "content_scripts": [
    {
      "matches": ["*://*.discogs.com/*"],
      "js": ["js/extension/user-preferences.js"],
      "css": ["css/dark-theme.css"],
      "run_at": "document_start"
    }
  ],

  "permissions": [
    "activeTab",
    "contextMenus",
    "storage",
    "scripting"
  ],

  "host_permissions": [
    "*://*.discogs.com/*"
  ],

  "web_accessible_resources": [
    {
      "resources": [
        "css/*",
        "js/extension/dependencies/*",
        "js/extension/features/*"
      ],
      "matches": ["*://*.discogs.com/*"]
    }
  ],

  "icons": {
    "16": "img/icon_16.png",
    "32": "img/icon_32.png",
    "48": "img/icon_48.png",
    "128": "img/icon_128.png"
  }
}
```

**Comment by salcido** [comment]
@Madslaw Thanks, there is still quite a bit of work to do on the extension to get it to pass Mozilla's code review. I'll need to remove all mention of Chrome from the documentation (shouldn't be too tedious). There's a lot of refactoring involved due to the `innerHTML` issues I mentioned above (unless Mozilla has changed their position on that in recent years). And I'll need to make some changes to Webpack in order to try and get it to render specific manifests for both Chrome and FF when the build command is run. 

If anyone is up for helping out with these tasks please feel free to submit a PR!

**Comment by Madslaw** [comment]
I don't know if this can help you save some time, but here are the results of the Mozilla's code review for my modified version of the 3.9.4 release (so just the manifest.json that is different). I had to go through this process to install the extension permanently in my browser. 

While it found 140 warning messages in the general tests (which most of them are the `innerHTML` issues) it still got the "Add-on passed validation" mention. I don't know if a public release gets more scrutinized than one marked as unlisted as I've never published a Firefox extensions. I'm guessing it may be best practice and safer to correct them regardless.

[Discogs Enhancer 3.9.4 - Firefox Code Review.pdf](https://github.com/user-attachments/files/20963099/Discogs.Enhancer.3.9.4.-.Firefox.Code.Review.pdf)

**Comment by salcido** [comment]
Thanks! I've got some time off coming up so I'll take another look at this.

**Comment by salcido** [comment]
Update: I've made some serious progress on this. I've updated Webpack so it will now generate Chrome and Firefox specific builds. There's new `watch` commands for Chrome, Firefox, and both Chrome & Firefox simultaneously. And a new manifest template for rendering Chrome/Firefox specific manifests. 

I realized there will likely be a need to migrate existing user configurations between browsers so the last thing I need to do is create a new feature that will allow users to backup / move their data. Unfortunately, because some data lives in discogs.com `localStorage` and some lives in `chrome.storage.sync`, users will likely need to follow a multi-step process to migrate their configurations. My goal is to keep the UI as simple as possible in order to minimize the chances of breaking the extension with malformed data. 

Once this is done, the final hurdle will be getting approval from Mozilla. I'm still worried about their previous objection to `innerHTML` usage. If that's still an issue I will likely ask for help with rewriting things since there is quite a bit of string templates used throughout the codebase.

**Comment by salcido** [comment]
Ok, I've merged the [Firefox Support PR](https://github.com/salcido/discogs-enhancer/pull/309). Next step is to submit it to Mozilla for review. I'll get that done this weekend and update this thread throughout the review process.

**Comment by nd-** [comment]
That's really great news \o/
And thanx !

**Comment by salcido** [comment]
It's been approved! 

https://addons.mozilla.org/en-US/firefox/addon/discogs-enhancer/

**Comment by jbyauk** [comment]
Dark mode is glorious, and there are even more features than I realized at first.  Thank you so much!
